### PR TITLE
[MCC-575482] Incorporate Protocol Test Suite and Implement

### PR DESCRIPTION
**Pull request by cmcinnes-mdsol**
@mdsol/team-16 this PR incorporate the new [mauth protocol test suite](https://github.com/mdsol/mauth-protocol-test-suite) into this client as a submodule. Opening as draft because the test suite PR has not yet been merged and because i still need to work on request path normalization. 

The PR adds some parsing code that turns the files in the test suite into runnable rspec tests.

The PR also implements (with unit tests) the new and improved MWSV2 protocol, including removing "double encoding" of query strings and adding some normalization of the `request_url_path`.

I also updated the contributing doc slightly as using submodules adds some complexity when cloning (i'd appreciate feedback on this section as i'm not super familiar with submodules).

I have not updated the changelog or version here. I'll do that in a follow up PR that makes the fallback to v1 configurable.

### CODE REVIEW

**Comment by cmcinnes-mdsol** [comment]
FYI, CI will fail until we make the mauth-protocol-test-suite public.

**Comment by cmcinnes-mdsol** [comment]
@mdsol/team-16 I initially added https://github.com/mdsol/mauth-protocol-test-suite as a git submodule and had it run on every travis build but had to opt for a file path approach (that doesn't run in travis) as the test suite repo is not yet public so can't be accessed from Travis here. 
To try out running the procotol test suite locally see the directions in `CONTRIBUTING.md`

**Comment by jfeltesse-mdsol** [comment]
@cmcinnes-mdsol any pressing reason we have to have that PR completed ASAP? I think making the submodules repo open source and leverage it here is the right thing to do. Have you made a request for open sourcing already or do you need help to do so?

**Comment by cmcinnes-mdsol** [comment]
Hey @jfeltesse-mdsol. I agree that open sourcing the test suite and using it as a submodule is the right way to go. I've obtained VP approval for open sourcing (from Sharlene Carnegie) but as @ejennings-mdsol pointed out we borrowed heavily from the aws test suite for the mauth protocol test suite and we need to figure out licensing with the legal team before we proceed. The aws test suite is licensed under the apache license (AFAIK we usually use the MIT license). 

We thought that might be a longer process so I went with the current approach instead to unblock this story.

File: spec/test_suite_parser.rb [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+      mc = ProtocolHelper::Config.mauth_client
+      mc.signed_headers_v2(req, time: ProtocolHelper::Config.request_time)
+    end
+
+    def write_authz
+      write_file('authz', JSON.pretty_generate(auth_headers))
+    end
+
+    def write_file(ext, contents)
+      File.open("#{CASE_PATH}/#{case_name}/#{case_name}.#{ext}", 'w') { |f| f.write(contents) }
```
Comment (jfeltesse-mdsol) [inline spec/test_suite_parser.rb:138]: `File.write("#{CASE_PATH}/#{case_name}/#{case_name}.#{ext}", contents)` ?

File: spec/signable_spec.rb [hunk 2]
```diff
@@ -204,4 +228,42 @@ class Dummy
       expect(dummy_inst.uri_escape(' ')).to eq('%20')
     end
   end
+
+  describe 'normalize_path' do
+    it 'normalizes self (".") in the path' do
+      path = '/./example/./.'
+      expected = '/example/'
+      expect(dummy_inst.normalize_path(path)).to eq(expected)
```
Comment (jfeltesse-mdsol) [inline spec/signable_spec.rb:236]: personally I would use shared_examples or even maybe, god forbid, "subject" but those specs are going to be refactored once the suite is cleared for open sourcing, right?

File: spec/signable_spec.rb [hunk 3]
```diff
@@ -204,4 +228,42 @@ class Dummy
...
+    end
+
+    it 'normalizes case of percent encoded characters' do
+      path = '/%2b'
+      expected = '/%2B'
+      expect(dummy_inst.normalize_path(path)).to eq(expected)
+    end
+
+    it 'normalizs multiple adjacent slashes to a single slash' do
```
Comment (jfeltesse-mdsol) [inline spec/signable_spec.rb:257]: typo

File: spec/signable_spec.rb (hunk 2 above)
Comment (cmcinnes-mdsol) [reply spec/signable_spec.rb:236]: I wasn't planning to refactor these specs after the suite is open sourced. What did you have in mind? The only thing change I had in mind once we open source is going back to including the test suite as a submodule like i had it in 71b249575c7942b8a93d9a40f6539f23bc2fba27

Comment (jfeltesse-mdsol) [reply spec/signable_spec.rb:236]: Might be that `subject` would be able to take your further but I can imagine something along the lines of the following to keep things simple:

```ruby
shared_examples_for "normalize_path" do |desc, path, expected|
  it "normalize_path: #{desc}" do
    expect(dummy_inst.normalize_path(path)).to eq(expected)
  end
end

include_examples 'normalize_path', 'self (".") in the path', '/./example/./.', '/example/'
include_examples 'normalize_path', 'parent ("..") in path', '/example/sample/..', '/example/'
# etc
```

Just my 2円, I don't think what you have here is bad, just a tad verbose IMHO.

Comment (cmcinnes-mdsol) [reply spec/signable_spec.rb:236]: ah yes sure thing, i'll update 👍

File: lib/mauth/request_and_response.rb [hunk 4]
```diff
@@ -78,10 +80,25 @@ def string_to_sign_v2(override_attrs)
...
+    # Addressable::URI.parse(path).normalize.to_s.squeeze('/')
+    def normalize_path(path)
+      return if path.nil?
+
+      # Addressable::URI.normalize_path normalizes `.` and `..` in path
+      #   i.e. /./example => /example ; /example/.. => /
+      # String#squeeze removes duplicated slahes i.e. /// => /
+      # String#gsub normalizes percent encoding to uppercase i.e. %cf%80 => %CF%80
+      Addressable::URI.normalize_path(path).squeeze('/').
+        gsub(/%[a-f0-9]{2}/) { |escape_sequence| escape_sequence.upcase }
```
[the remark this answers is not in the document]
Comment (cmcinnes-mdsol) [reply lib/mauth/request_and_response.rb:92]: nice, will update

**Pull request merged**